[Filebeat] Adds oauth2 support for httpjson input

marc-gr opened the pull request:
## What does this PR do?

Adds oauth2 support to filebeat HTTPJSON input.

## Why is it important?

This will let us pull data from oauth2 authenticated sources, like AzureAD or GSuite.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
~I have made corresponding change to the default configuration files~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

Closes #18415

<!-- Recommended
## Use cases

Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

elasticmachine commented:
Pinging @elastic/siem (Team:SIEM)

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-18892/detail/PR-18892/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-18892/detail/PR-18892/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-18892/detail/PR-18892/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-18892/detail/PR-18892/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_18892.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #18892 opened]`

* Start Time: 2020-06-02T10:07:30.902+0000

* Duration: 47 min 0 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 2210  |
| Skipped      | 382 |
| Total        | 2592   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->

P1llus commented on x-pack/filebeat/input/httpjson/config.go:
```diff
@@ -84,6 +94,11 @@ func (c *config) Validate() error {
 			}
 		}
 	}
+	if c.OAuth2 != nil {
```
Maybe others might think differently, but maybe nice to have a enable/disable flag like pagination has, that way they could easily disable it on the fly without having to remove all configuration

P1llus replied:
If we are still looking for this being nil or not, then having enabled/disable option will still throw an error I believe? Maybe add a second condition saying "if its not null and not disabled"?

marc-gr replied:
@andrewkroh pointed out the same, WDYT about this? https://github.com/elastic/beats/pull/18892#discussion_r435727095

P1llus commented on x-pack/filebeat/input/httpjson/input.go:
```diff
@@ -462,6 +444,47 @@ func (in *HttpjsonInput) Wait() {
+	creds := clientcredentials.Config{
+		ClientID:       in.config.OAuth2.ClientID,
+		ClientSecret:   in.config.OAuth2.ClientSecret,
+		TokenURL:       in.config.OAuth2.TokenURL,
```
Did we want later to just add it so that tenant-id or client-id would be appended to the URL, in a case we might have a list of endpoints pre-made, that way the user would not need to know what to configure here.

andrewkroh commented on x-pack/filebeat/input/httpjson/config.go:
```diff
@@ -84,6 +83,14 @@ func (c *config) Validate() error {
 			}
 		}
 	}
+	if c.OAuth2 != nil {
```
```suggestion
	if c.OAuth2.IsEnabled() {
```

marc-gr replied:
Here I followed the same pattern as `Pagination`, which checks for it being enabled at the moment of using it, but still validates its config. To me it kind of made sense since even if disabled a user needs to introduce a valid config for it. Otherwise I would change also how `Pagination` is handled just to be consistent here and check for `IsEnabled` in both config validation and usage. WDYT?

andrewkroh replied:
I think using `IsEnabled()` is correct now that you've refactored it a bit (before you we're correct). For example it would be invalid to use `api_key` or `authentication` only when oauth is enabled.

marc-gr replied:
Good point! :+1:

andrewkroh commented on x-pack/filebeat/input/httpjson/config.go:
```diff
@@ -84,6 +83,14 @@ func (c *config) Validate() error {
 	}
+	if c.OAuth2 != nil {
+		if c.APIKey != "" || c.AuthenticationScheme != "" {
+			return errors.Errorf("invalid configuration: both oauth2 and api_key or authentication_scheme cannot be set simultaneously")
```
```suggestion
			return errors.Errorf("invalid configuration: oauth2 and api_key or authentication_scheme cannot be set simultaneously")
```

andrewkroh commented on x-pack/filebeat/input/httpjson/config.go:
```diff
@@ -84,6 +83,14 @@ func (c *config) Validate() error {
+		if c.APIKey != "" || c.AuthenticationScheme != "" {
+			return errors.Errorf("invalid configuration: both oauth2 and api_key or authentication_scheme cannot be set simultaneously")
+		}
+		if err := c.OAuth2.validate(); err != nil {
```
I think you can move the `OAuth2` validation into its own `Validate() error` function and go-ucfg should automatically invoke it when unpacking the config.

andrewkroh commented on x-pack/filebeat/input/httpjson/config_oauth.go:
```diff
@@ -0,0 +1,161 @@
+// OAuth2 contains information about oauth2 authentication settings.
+type OAuth2 struct {
+	// common oauth fields
+	ClientID       string              `config:"client_id"`
```
We tend to use hierarchical config parameters in Beats. Like group `id` and `secret` under `client`.
```suggestion
	ClientID       string              `config:"client.id"`
```

andrewkroh commented on x-pack/filebeat/input/httpjson/config_oauth.go:
```diff
@@ -0,0 +1,161 @@
+	TokenURL       string              `config:"token_url"`
+
+	// google specific
+	GoogleCredentialsFile string `config:"google_credentials_file"`
+	GoogleCredentialsJSON []byte `config:"google_credentials_json"`
+	GoogleJWTFile         string `config:"google_jwt_file"`
```
```suggestion
	GoogleCredentialsFile string `config:"google.credentials_file"`
	GoogleCredentialsJSON []byte `config:"google.credentials_json"`
	GoogleJWTFile         string `config:"google.jwt_file"`
```

or this since I don't remember if embedding a dot in the config tag works (try it)

```suggestion
	Google struct {
	    CredentialsFile string `config:"credentials_file"`
	    ...
	} `config:"google"`
```

andrewkroh commented on x-pack/filebeat/input/httpjson/config_oauth.go:
```diff
@@ -0,0 +1,161 @@
+	GoogleJWTFile         string `config:"google_jwt_file"`
+
+	// microsoft azure specific
+	AzureTenantID string `config:"azure_tenant_id"`
```
```suggestion
	AzureTenantID string `config:"azure.tenant_id"`
```

(to be consistent with `google.X`)

andrewkroh commented on x-pack/filebeat/input/httpjson/config_oauth.go:
```diff
@@ -0,0 +1,161 @@
+func (o *OAuth2) Client(ctx context.Context, client *http.Client) (*http.Client, error) {
+	ctx = context.WithValue(ctx, oauth2.HTTPClient, client)
+
+	if o.Provider != OAuth2ProviderGoogle {
```
I think a `switch o.Provider` statement would help make this section more clear that all the provider cases are explicitly handled.

P1llus commented on x-pack/filebeat/input/httpjson/config.go:
```diff
@@ -62,9 +63,7 @@ type RateLimit struct {
-	case "GET":
-		break
-	case "POST":
+	case "GET", "POST":
```
Could you add validation for any raw json strings in your config? Example from my http_endpoint input:

```
func (c *config) Validate() error {
	if !json.Valid([]byte(c.ResponseBody)) {
		return errors.New("response_body must be valid JSON")
	}

	return nil
}
```

marc-gr replied:
Added it in the places where it is expected: https://github.com/elastic/beats/pull/18892/files#diff-e3d27d33f878a5fb0a720224c80ad7aaR135 and https://github.com/elastic/beats/pull/18892/files#diff-e3d27d33f878a5fb0a720224c80ad7aaR171

andrewkroh commented on x-pack/filebeat/docs/inputs/input-httpjson.asciidoc:
```diff
@@ -249,6 +280,117 @@ information.
+combination with it. It is not required.
+
+NOTE: For information about where to find it, you can refer to
+https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal
```
```suggestion
https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal.
```

andrewkroh commented on x-pack/filebeat/input/httpjson/config_test.go:
```diff
@@ -0,0 +1,418 @@
+	}
+}
+
+func TestConfigValidationCase14(t *testing.T) {
```
I'm really happy to have all the new test cases. I'm not keen on the numerical naming pattern that got established here. Maybe break for this pattern for the new tests. I'd prefer something that describes the requirements/assertion being made about the code (e.g. `TestConfigValidateRequiresCreds` or use `TestConfigValidate` with `t.Run("validate requires credentials", func(...)`).

marc-gr commented:
Moved the config tests from `TestConfigValidateNN` to a table test. Also changed `config.Validate` to use `IsEnabled` and added an extra test for it.  Did some changes on the docs, please tell me if you think they need any more 👍
